File: frontend/src/api/pmo.ts
// PMO API — P1: расписание (Гантт) + зависимости задач.
import { api } from './client';

export type ScheduleBarKind = 'project' | 'task';
export type DepType = 'FS' | 'SS' | 'FF' | 'SF';

export interface ScheduleBar {
  id: string;
  kind: ScheduleBarKind;
  project_id: string | null;
  title: string;
  status: string;
  progress_percent: number;
  start: string | null;          // YYYY-MM-DD
  due: string | null;
  baseline_start: string | null;
  baseline_due: string | null;
  is_milestone: boolean;
  assignee_name: string | null;
  direction: string | null;
  slip_days: number;
  on_critical_path: boolean;
  predecessor_ids: string[];
  blocked: boolean;
}

export interface ScheduleResponse {
  company_code: string;
  year: number | null;
  as_of: string;
  bars: ScheduleBar[];
  portfolio_slip_days: number;
  forecast_finish: string | null;
  baseline_finish: string | null;
  critical_path_ids: string[];
  overdue_count: number;
  blocked_count: number;
}

export interface DependencyRead {
  id: string;
  predecessor_id: string;
  successor_id: string;
  dep_type: string;
  lag_days: number;
}

// ── P2: RAID / Health / Status reports ─────────────────────────────────
export type RaidKind = "risk" | "assumption" | "issue" | "dependency";
export type RaidSeverity = "low" | "medium" | "high" | "critical";
export type RaidStatus = "open" | "mitigating" | "closed";
export type RaidPolarity = "threat" | "opportunity";
export type Engagement = "unaware" | "resistant" | "neutral" | "supportive" | "leading";

export interface RaidItem {
  id: string;
  company_id: string | null;
  project_id: string | null;
  kind: RaidKind;
  title: string;
  description: string | null;
  owner_id: string | null;
  owner_name: string | null;
  severity: RaidSeverity;
  probability: number;
  impact: number;
  score: number;
  polarity: RaidPolarity;
  response_strategy: string | null;
  status: RaidStatus;
  mitigation: string | null;
  due_date: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RaidPayload {
  kind?: RaidKind;
  title: string;
  description?: string | null;
  project_id?: string | null;
  owner_name?: string | null;
  severity?: RaidSeverity;
  probability?: number;
  impact?: number;
  polarity?: RaidPolarity;
  response_strategy?: string | null;
  status?: RaidStatus;
  mitigation?: string | null;
  due_date?: string | null;
}

export interface Stakeholder {
  id: string;
  company_id: string | null;
  project_id: string | null;
  name: string;
  role: string | null;
  organization: string | null;
  power: number;
  interest: number;
  engagement_current: Engagement;
  engagement_desired: Engagement;
  strategy: string | null;
  contact: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface StakeholderPayload {
  name: string;
  role?: string | null;
  organization?: string | null;
  power?: number;
  interest?: number;
  engagement_current?: Engagement;
  engagement_desired?: Engagement;
  strategy?: string | null;
  contact?: string | null;
  notes?: string | null;
}

// ── Журнал: уроки + изменения ──
export type LessonKind = "success" | "problem" | "recommendation";
export interface Lesson {
  id: string;
  company_id: string | null;
  project_id: string | null;
  kind: LessonKind;
  title: string;
  description: string | null;
  recommendation: string | null;
  owner_name: string | null;
  created_at: string;
  updated_at: string;
}
export interface LessonPayload {
  kind?: LessonKind;
  title: string;
  description?: string | null;
  recommendation?: string | null;
  owner_name?: string | null;
}

export type ChangeKind = "scope" | "schedule" | "cost" | "quality" | "other";
export type ChangeStatus = "proposed" | "approved" | "rejected" | "implemented";
export interface ChangeItem {
  id: string;
  company_id: string | null;
  project_id: string | null;
  kind: ChangeKind;
  title: string;
  description: string | null;
  impact: string | null;
  requested_by: string | null;
  status: ChangeStatus;
  decided_by: string | null;
  decided_at: string | null;
  created_at: string;
  updated_at: string;
}
export interface ChangePayload {
  kind?: ChangeKind;
  title: string;
  description?: string | null;
  impact?: string | null;
  requested_by?: string | null;
  status?: ChangeStatus;
  decided_by?: string | null;
}

export interface HealthProject {
  project_id: string | null;
  title: string;
  rag: "green" | "amber" | "red";
  progress_percent: number;
  slip_days: number;
  overdue_count: number;
  blocked_count: number;
  open_risks: number;
  high_risks: number;
  reasons: string[];
}

export interface HealthResponse {
  company_code: string;
  as_of: string;
  portfolio_rag: "green" | "amber" | "red";
  projects: HealthProject[];
  green: number;
  amber: number;
  red: number;
  open_risks: number;
  high_risks: number;
}

export interface StatusReport {
  id: string;
  company_id: string | null;
  project_id: string | null;
  period: string | null;
  rag: "green" | "amber" | "red";
  summary: string | null;
  metrics: any | null;
  created_at: string;
}

export const pmoApi = {
  getSchedule(code: string, year?: number | null): Promise<ScheduleResponse> {
    const q = year != null ? `?year=${year}` : '';
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/schedule${q}`).then(r => r.data);
  },
  listDependencies(code: string): Promise<DependencyRead[]> {
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/dependencies`).then(r => r.data);
  },
  createDependency(body: {
    predecessor_id: string;
    successor_id: string;
    dep_type?: DepType;
    lag_days?: number;
  }): Promise<DependencyRead> {
    return api.post('/pmo/dependencies', body).then(r => r.data);
  },
  deleteDependency(id: string): Promise<void> {
    return api.delete(`/pmo/dependencies/${id}`).then(() => undefined);
  },

  // ── RAID ──
  listRaid(code: string, opts?: { kind?: string; status?: string }): Promise<RaidItem[]> {
    const p = new URLSearchParams();
    if (opts?.kind) p.set("kind", opts.kind);
    if (opts?.status) p.set("status_filter", opts.status);
    const q = p.toString() ? `?${p.toString()}` : "";
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/raid${q}`).then(r => r.data);
  },
  createRaid(code: string, body: RaidPayload): Promise<RaidItem> {
    return api.post(`/pmo/companies/${encodeURIComponent(code)}/raid`, body).then(r => r.data);
  },
  updateRaid(id: string, body: Partial<RaidPayload>): Promise<RaidItem> {
    return api.patch(`/pmo/raid/${id}`, body).then(r => r.data);
  },
  deleteRaid(id: string): Promise<void> {
    return api.delete(`/pmo/raid/${id}`).then(() => undefined);
  },

  // ── Health / Status reports ──
  getHealth(code: string): Promise<HealthResponse> {
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/health`).then(r => r.data);
  },
  listStatusReports(code: string): Promise<StatusReport[]> {
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/status-reports`).then(r => r.data);
  },
  createStatusReport(code: string, body: { project_id?: string | null; use_ai?: boolean }): Promise<StatusReport> {
    return api.post(`/pmo/companies/${encodeURIComponent(code)}/status-reports`, body).then(r => r.data);
  },

  // ── Стейкхолдеры ──
  listStakeholders(code: string): Promise<Stakeholder[]> {
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/stakeholders`).then(r => r.data);
  },
  createStakeholder(code: string, body: StakeholderPayload): Promise<Stakeholder> {
    return api.post(`/pmo/companies/${encodeURIComponent(code)}/stakeholders`, body).then(r => r.data);
  },
  updateStakeholder(id: string, body: Partial<StakeholderPayload>): Promise<Stakeholder> {
    return api.patch(`/pmo/stakeholders/${id}`, body).then(r => r.data);
  },
  deleteStakeholder(id: string): Promise<void> {
    return api.delete(`/pmo/stakeholders/${id}`).then(() => undefined);
  },

  // ── Журнал: уроки ──
  listLessons(code: string): Promise<Lesson[]> {
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/lessons`).then(r => r.data);
  },
  createLesson(code: string, body: LessonPayload): Promise<Lesson> {
    return api.post(`/pmo/companies/${encodeURIComponent(code)}/lessons`, body).then(r => r.data);
  },
  updateLesson(id: string, body: Partial<LessonPayload>): Promise<Lesson> {
    return api.patch(`/pmo/lessons/${id}`, body).then(r => r.data);
  },
  deleteLesson(id: string): Promise<void> {
    return api.delete(`/pmo/lessons/${id}`).then(() => undefined);
  },

  // ── Журнал: изменения ──
  listChanges(code: string): Promise<ChangeItem[]> {
    return api.get(`/pmo/companies/${encodeURIComponent(code)}/changes`).then(r => r.data);
  },
  createChange(code: string, body: ChangePayload): Promise<ChangeItem> {
    return api.post(`/pmo/companies/${encodeURIComponent(code)}/changes`, body).then(r => r.data);
  },
  updateChange(id: string, body: Partial<ChangePayload>): Promise<ChangeItem> {
    return api.patch(`/pmo/changes/${id}`, body).then(r => r.data);
  },
  deleteChange(id: string): Promise<void> {
    return api.delete(`/pmo/changes/${id}`).then(() => undefined);
  },
};

File: backend/app/schemas/pmo.py
"""PMO schemas — P1: расписание / Гантт / зависимости.

Schedule = плоский список баров (проекты + их задачи) с базовым планом,
слипом, флагом критического пути и списком предшественников. Фронт рендерит
таймлайн из этого DTO; критический путь и слип считаются на бэкенде.
"""
from datetime import date, datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class ScheduleBar(BaseModel):
    """Один бар на таймлайне — проект или задача."""
    id: UUID
    kind: Literal["project", "task"]
    project_id: Optional[UUID] = None     # для задач — родитель (группировка строк)
    title: str
    status: str
    progress_percent: int = 0

    start: Optional[date] = None
    due: Optional[date] = None
    baseline_start: Optional[date] = None
    baseline_due: Optional[date] = None

    is_milestone: bool = False
    assignee_name: Optional[str] = None
    direction: Optional[str] = None

    # Расчётные (бэкенд):
    slip_days: int = 0                    # due − baseline_due (>0 = опоздание)
    on_critical_path: bool = False
    predecessor_ids: list[UUID] = Field(default_factory=list)
    blocked: bool = False                 # есть незавершённый предшественник


class ScheduleResponse(BaseModel):
    company_code: str
    year: Optional[int] = None
    as_of: date
    bars: list[ScheduleBar] = Field(default_factory=list)
    # Сводка портфеля:
    portfolio_slip_days: int = 0          # макс. слип по критическому пути
    forecast_finish: Optional[date] = None
    baseline_finish: Optional[date] = None
    critical_path_ids: list[UUID] = Field(default_factory=list)
    overdue_count: int = 0
    blocked_count: int = 0


# ─── Dependencies ──────────────────────────────────────────────────────

DepType = Literal["FS", "SS", "FF", "SF"]


class DependencyCreate(BaseModel):
    predecessor_id: UUID
    successor_id: UUID
    dep_type: DepType = "FS"
    lag_days: int = 0


class DependencyRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    predecessor_id: UUID
    successor_id: UUID
    dep_type: str
    lag_days: int


# ─── P2: RAID-реестр ───────────────────────────────────────────────────

RaidKind = Literal["risk", "assumption", "issue", "dependency"]
RaidSeverity = Literal["low", "medium", "high", "critical"]
RaidStatus = Literal["open", "mitigating", "closed"]


class RaidItemRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    kind: str
    title: str
    description: Optional[str] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = None
    severity: str
    probability: int
    impact: int
    score: int
    polarity: str = "threat"
    response_strategy: Optional[str] = None
    status: str
    mitigation: Optional[str] = None
    due_date: Optional[date] = None
    closed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class RaidItemCreate(BaseModel):
    kind: RaidKind = "risk"
    title: str = Field(..., min_length=1, max_length=512)
    description: Optional[str] = None
    project_id: Optional[UUID] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = Field(None, max_length=255)
    severity: RaidSeverity = "medium"
    probability: int = Field(3, ge=1, le=5)
    impact: int = Field(3, ge=1, le=5)
    polarity: Literal["threat", "opportunity"] = "threat"
    response_strategy: Optional[str] = Field(None, max_length=16)
    status: RaidStatus = "open"
    mitigation: Optional[str] = None
    due_date: Optional[date] = None


class RaidItemUpdate(BaseModel):
    kind: Optional[RaidKind] = None
    title: Optional[str] = Field(None, min_length=1, max_length=512)
    description: Optional[str] = None
    project_id: Optional[UUID] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = None
    severity: Optional[RaidSeverity] = None
    probability: Optional[int] = Field(None, ge=1, le=5)
    impact: Optional[int] = Field(None, ge=1, le=5)
    polarity: Optional[Literal["threat", "opportunity"]] = None
    response_strategy: Optional[str] = Field(None, max_length=16)
    status: Optional[RaidStatus] = None
    mitigation: Optional[str] = None
    due_date: Optional[date] = None


# ─── PMBOK 7: Стейкхолдеры ─────────────────────────────────────────────

Engagement = Literal["unaware", "resistant", "neutral", "supportive", "leading"]


class StakeholderRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    name: str
    role: Optional[str] = None
    organization: Optional[str] = None
    power: int
    interest: int
    engagement_current: str
    engagement_desired: str
    strategy: Optional[str] = None
    contact: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class StakeholderCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    role: Optional[str] = Field(None, max_length=255)
    organization: Optional[str] = Field(None, max_length=255)
    project_id: Optional[UUID] = None
    power: int = Field(3, ge=1, le=5)
    interest: int = Field(3, ge=1, le=5)
    engagement_current: Engagement = "neutral"
    engagement_desired: Engagement = "supportive"
    strategy: Optional[str] = None
    contact: Optional[str] = Field(None, max_length=255)
    notes: Optional[str] = None


class StakeholderUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    role: Optional[str] = Field(None, max_length=255)
    organization: Optional[str] = Field(None, max_length=255)
    project_id: Optional[UUID] = None
    power: Optional[int] = Field(None, ge=1, le=5)
    interest: Optional[int] = Field(None, ge=1, le=5)
    engagement_current: Optional[Engagement] = None
    engagement_desired: Optional[Engagement] = None
    strategy: Optional[str] = None
    contact: Optional[str] = Field(None, max_length=255)
    notes: Optional[str] = None


# ─── PMBOK 7: Журнал — извлечённые уроки + изменения ───────────────────

LessonKind = Literal["success", "problem", "recommendation"]


class LessonRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    kind: str
    title: str
    description: Optional[str] = None
    recommendation: Optional[str] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class LessonCreate(BaseModel):
    kind: LessonKind = "recommendation"
    title: str = Field(..., min_length=1, max_length=512)
    description: Optional[str] = None
    recommendation: Optional[str] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = Field(None, max_length=255)
    project_id: Optional[UUID] = None


class LessonUpdate(BaseModel):
    kind: Optional[LessonKind] = None
    title: Optional[str] = Field(None, min_length=1, max_length=512)
    description: Optional[str] = None
    recommendation: Optional[str] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = None
    project_id: Optional[UUID] = None


ChangeKind = Literal["scope", "schedule", "cost", "quality", "other"]
ChangeStatus = Literal["proposed", "approved", "rejected", "implemented"]


class ChangeRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    kind: str
    title: str
    description: Optional[str] = None
    impact: Optional[str] = None
    requested_by: Optional[str] = None
    status: str
    decided_by: Optional[str] = None
    decided_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ChangeCreate(BaseModel):
    kind: ChangeKind = "scope"
    title: str = Field(..., min_length=1, max_length=512)
    description: Optional[str] = None
    impact: Optional[str] = None
    requested_by: Optional[str] = Field(None, max_length=255)
    status: ChangeStatus = "proposed"
    decided_by: Optional[str] = Field(None, max_length=255)
    project_id: Optional[UUID] = None


class ChangeUpdate(BaseModel):
    kind: Optional[ChangeKind] = None
    title: Optional[str] = Field(None, min_length=1, max_length=512)
    description: Optional[str] = None
    impact: Optional[str] = None
    requested_by: Optional[str] = None
    status: Optional[ChangeStatus] = None
    decided_by: Optional[str] = None
    project_id: Optional[UUID] = None


# ─── P3: Устав проекта (Charter) ───────────────────────────────────────

CharterStatus = Literal["draft", "approved"]


class CharterRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    project_title: Optional[str] = None
    purpose: Optional[str] = None
    objectives: Optional[str] = None
    scope_in: Optional[str] = None
    scope_out: Optional[str] = None
    success_criteria: Optional[str] = None
    deliverables: Optional[str] = None
    milestones: Optional[str] = None
    assumptions: Optional[str] = None
    constraints: Optional[str] = None
    sponsor_name: Optional[str] = None
    manager_name: Optional[str] = None
    budget_amount: Optional[float] = None
    start_date: Optional[date] = None
    target_end_date: Optional[date] = None
    status: str = "draft"
    approved_by: Optional[str] = None
    approved_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class CharterCreate(BaseModel):
    project_id: Optional[UUID] = None
    project_title: Optional[str] = Field(None, max_length=512)
    purpose: Optional[str] = None
    objectives: Optional[str] = None
    scope_in: Optional[str] = None
    scope_out: Optional[str] = None
    success_criteria: Optional[str] = None
    deliverables: Optional[str] = None
    milestones: Optional[str] = None
    assumptions: Optional[str] = None
    constraints: Optional[str] = None
    sponsor_name: Optional[str] = Field(None, max_length=255)
    manager_name: Optional[str] = Field(None, max_length=255)
    budget_amount: Optional[float] = None
    start_date: Optional[date] = None
    target_end_date: Optional[date] = None


class CharterUpdate(BaseModel):
    project_id: Optional[UUID] = None
    project_title: Optional[str] = Field(None, max_length=512)
    purpose: Optional[str] = None
    objectives: Optional[str] = None
    scope_in: Optional[str] = None
    scope_out: Optional[str] = None
    success_criteria: Optional[str] = None
    deliverables: Optional[str] = None
    milestones: Optional[str] = None
    assumptions: Optional[str] = None
    constraints: Optional[str] = None
    sponsor_name: Optional[str] = Field(None, max_length=255)
    manager_name: Optional[str] = Field(None, max_length=255)
    budget_amount: Optional[float] = None
    start_date: Optional[date] = None
    target_end_date: Optional[date] = None
    status: Optional[CharterStatus] = None


# ─── P3: Освоенный объём (EVM) ─────────────────────────────────────────

class EvmProject(BaseModel):
    project_id: Optional[UUID] = None
    title: str
    progress_percent: int = 0
    planned_percent: Optional[int] = None
    bac: Optional[float] = None    # Budget At Completion
    ev: Optional[float] = None     # Earned Value
    pv: Optional[float] = None     # Planned Value
    ac: Optional[float] = None     # Actual Cost
    spi: Optional[float] = None    # Schedule Performance Index
    cpi: Optional[float] = None    # Cost Performance Index
    sv: Optional[float] = None     # Schedule Variance
    cv: Optional[float] = None     # Cost Variance
    eac: Optional[float] = None    # Estimate At Completion
    etc: Optional[float] = None    # Estimate To Complete
    vac: Optional[float] = None    # Variance At Completion
    tcpi: Optional[float] = None   # To-Complete Performance Index
    rag: str = "na"                # green | amber | red | na


class EvmResponse(BaseModel):
    company_code: str
    as_of: date
    bac: Optional[float] = None
    ev: Optional[float] = None
    pv: Optional[float] = None
    ac: Optional[float] = None
    spi: Optional[float] = None
    cpi: Optional[float] = None
    sv: Optional[float] = None
    cv: Optional[float] = None
    eac: Optional[float] = None
    etc: Optional[float] = None
    vac: Optional[float] = None
    tcpi: Optional[float] = None
    rag: str = "na"
    projects: list[EvmProject] = Field(default_factory=list)
    budgeted_count: int = 0
    total_count: int = 0


# ─── P3: Команда / загрузка (Workload) ─────────────────────────────────

class WorkloadPerson(BaseModel):
    person_id: Optional[UUID] = None
    name: str
    assigned: int = 0
    open: int = 0
    overdue: int = 0
    done: int = 0
    load: int = 0
    capacity: str = "free"   # free | normal | high | overload


class WorkloadResponse(BaseModel):
    company_code: str
    as_of: date
    people: list[WorkloadPerson] = Field(default_factory=list)
    total_people: int = 0
    total_open: int = 0
    unassigned_open: int = 0
    max_load: int = 0


# ─── P3: RACI-матрица ──────────────────────────────────────────────────

RaciRole = Literal["R", "A", "C", "I"]


class RaciRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    item_label: str
    person_name: str
    person_id: Optional[UUID] = None
    role: str
    created_at: datetime
    updated_at: datetime


class RaciCreate(BaseModel):
    item_label: str = Field(..., min_length=1, max_length=512)
    person_name: str = Field(..., min_length=1, max_length=255)
    person_id: Optional[UUID] = None
    role: RaciRole = "R"
    project_id: Optional[UUID] = None


class RaciUpdate(BaseModel):
    item_label: Optional[str] = Field(None, min_length=1, max_length=512)
    person_name: Optional[str] = Field(None, min_length=1, max_length=255)
    person_id: Optional[UUID] = None
    role: Optional[RaciRole] = None
    project_id: Optional[UUID] = None


# ─── P3: Agile / спринты (спринт группирует существующие задачи) ────────

SprintStatus = Literal["planned", "active", "done"]


class SprintRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    name: str
    goal: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: str
    capacity_points: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class SprintCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    goal: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: SprintStatus = "planned"
    capacity_points: Optional[int] = None
    project_id: Optional[UUID] = None


class SprintUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    goal: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: Optional[SprintStatus] = None
    capacity_points: Optional[int] = None
    project_id: Optional[UUID] = None


class AgileTask(BaseModel):
    """Задача в Agile-представлении (бэклог/доска спринта)."""
    id: UUID
    title: str
    status: str
    priority: str = "medium"
    tags: list[str] = Field(default_factory=list)
    project_id: Optional[UUID] = None
    project_title: Optional[str] = None
    assignee_id: Optional[UUID] = None
    assignee_name: Optional[str] = None
    story_points: Optional[int] = None
    sprint_id: Optional[UUID] = None
    due_date: Optional[date] = None
    weight: int = 1


class AgileResponse(BaseModel):
    company_code: str
    sprints: list[SprintRead] = Field(default_factory=list)
    tasks: list[AgileTask] = Field(default_factory=list)
    backlog_count: int = 0


class TaskAgilePatch(BaseModel):
    """Привязка/правка задачи в Agile: спринт, story points, статус (drag)."""
    sprint_id: Optional[UUID] = None
    story_points: Optional[int] = None
    status: Optional[str] = None


# ─── P2: Здоровье (авто-RAG) ───────────────────────────────────────────

class HealthProject(BaseModel):
    project_id: Optional[UUID] = None
    title: str
    rag: str                    # green | amber | red
    progress_percent: int = 0
    slip_days: int = 0
    overdue_count: int = 0
    blocked_count: int = 0
    open_risks: int = 0
    high_risks: int = 0
    reasons: list[str] = Field(default_factory=list)


class HealthResponse(BaseModel):
    company_code: str
    as_of: date
    portfolio_rag: str
    projects: list[HealthProject] = Field(default_factory=list)
    green: int = 0
    amber: int = 0
    red: int = 0
    open_risks: int = 0
    high_risks: int = 0


# ─── P2: Статус-отчёты ─────────────────────────────────────────────────

class StatusReportRead(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    company_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    period: Optional[date] = None
    rag: str
    summary: Optional[str] = None
    metrics: Optional[dict] = None
    created_at: datetime


class StatusReportCreate(BaseModel):
    project_id: Optional[UUID] = None   # None = портфельный отчёт
    use_ai: bool = False                # AI-резюме, если ИИ-движок доступен
